python/vast/utils/arrow.py

```diff
@@ -1,7 +1,125 @@
-import ipaddress
+import json
+import ipaddress as ip
+from typing import SupportsBytes
 
 import pyarrow as pa
 
+# Accepts a 128-bit buffer holding an IPv6 address and returns an IPv4 or IPv6
+# address.
+def unpack_ip(buffer: SupportsBytes) -> ip.IPv4Address | ip.IPv6Address:
+    num = int.from_bytes(buffer, byteorder="big")
+    # Convert IPv4 mapped addresses back to regular IPv4.
+    # https://tools.ietf.org/html/rfc4291#section-2.5.5.2
+    if (num >> 32) == 65535:
+        num = num - (65535 << 32)
+    return ip.ip_address(num)
+
+
+class AddressScalar(pa.ExtensionScalar):
+    def as_py(self) -> ip.IPv4Address | ip.IPv6Address:
+        return unpack_ip(self.value.as_py())
+
+
+class AddressType(pa.ExtensionType):
+    ext_name = "vast.address_type"
+    ext_type = pa.binary(16)
+
+    def __init__(self):
+        pa.ExtensionType.__init__(self, self.ext_type, self.ext_name)
+
+    def __arrow_ext_serialize__(self) -> bytes:
+        return self.ext_name.encode()
+
+    @classmethod
+    def __arrow_ext_deserialize__(self, storage_type, serialized: bytes):
+        if serialized.decode() != self.ext_name:
+            raise TypeError("type identifier does not match")
+        if storage_type != self.ext_type:
+            raise TypeError("storage type does not match")
+
+    def __reduce__(self):
+        return AddressScalar, ()
+
+    def __arrow_ext_scalar_class__(self):
+        return AddressScalar
+
+
+class SubnetScalar(pa.ExtensionScalar):
+    def as_py(self) -> ip.IPv4Network | ip.IPv6Network:
+        address = unpack_ip(self.value[0].as_py())
+        length = self.value[1].as_py()
+        return ip.ip_network((address, length), strict=False)
+
+
+class SubnetType(pa.ExtensionType):
+    ext_name = "vast.subnet_type"
+    # This gives us a pyarrow.lib.ArrowNotImplementedError :-/
+    # ext_type = pa.struct([("address", AddressType()),
+    #                       ("length", pa.uint8())])
+    ext_type = pa.struct([("address", pa.binary(16)), ("length", pa.uint8())])
+
+    def __init__(self):
+        pa.ExtensionType.__init__(self, self.ext_type, self.ext_name)
+
+    def __arrow_ext_serialize__(self) -> bytes:
+        return self.ext_name.encode()
+
+    @classmethod
+    def __arrow_ext_deserialize__(self, storage_type, serialized: bytes):
+        if serialized.decode() != self.ext_name:
+            raise TypeError("type identifier does not match")
+        if storage_type != self.ext_type:
+            raise TypeError("storage type does not match")
+        return SubnetType()
+
+    def __reduce__(self):
+        return SubnetScalar, ()
+
+    def __arrow_ext_scalar_class__(self):
+        return SubnetScalar
+
+
+class EnumScalar(pa.ExtensionScalar):
+    def as_py(self) -> str:
+        return self.type.field(self.value.as_py())
+
+
+class EnumType(pa.ExtensionType):
+    ext_name = "vast.enumeration_type"
```

Comment:
```suggestion
    ext_name = "vast.enumeration"
```

https://github.com/tenzir/vast/blob/1c3eec37e96db254ad73a405164718b4d435d779/libvast/include/vast/type.hpp#L1015-L1016